Source/WebCore/page/ElementTargetingController.cpp

```diff
@@ -860,6 +861,129 @@ Vector<TargetedElementInfo> ElementTargetingController::findTargets(TargetedElem
+
+    processingIDs.remove(currentElementID);
+    depthSortedIDs.append(currentElementID);
+}
+
+Vector<WebCore::ElementIdentifier> ElementTargetingController::topologicallySortElements(const HashMap<WebCore::ElementIdentifier, HashSet<WebCore::ElementIdentifier>> &elementIDToOccludedElementIDs)
+{
+    Vector<WebCore::ElementIdentifier> depthSortedIDs;
+    HashSet<WebCore::ElementIdentifier> processingIDs;
+    HashSet<WebCore::ElementIdentifier> unprocessedIDs;
+
+    const auto elementIDs = elementIDToOccludedElementIDs.keys();
+    unprocessedIDs.add(elementIDs.begin(), elementIDs.end());
+
+    while (!unprocessedIDs.isEmpty() || !processingIDs.isEmpty()) {
+        if (unprocessedIDs.isEmpty()) {
+            ASSERT_NOT_REACHED();
+            break;
+        }
+
+        topologicallySortElementsHelper(*unprocessedIDs.begin(), depthSortedIDs, processingIDs, unprocessedIDs, elementIDToOccludedElementIDs);
+    }
+
+    depthSortedIDs.reverse();
+    return depthSortedIDs;
+}
+
+Vector<Vector<TargetedElementInfo>> ElementTargetingController::findAllTargets(WebCore::FloatSize minimumSize)
+{
+    RefPtr page = m_page.get();
+    if (!page) {
+        ASSERT_NOT_REACHED();
+        return { };
+    }
+
+    RefPtr mainFrame = dynamicDowncast<LocalFrame>(page->mainFrame());
+    if (!mainFrame) {
+        ASSERT_NOT_REACHED();
+        return { };
+    }
+
+    RefPtr view = mainFrame->view();
+    if (!view) {
+        ASSERT_NOT_REACHED();
+        return { };
+    }
+
+    const auto viewportRect = view->unobscuredContentRect();
+    const auto hitTestRegionWidth = minimumSize.width();
+    const auto hitTestRegionHeight = minimumSize.height();
+
+    Vector<Vector<TargetedElementInfo>> targetsList;
+    for (auto x = viewportRect.x() + std::floor(hitTestRegionWidth / 2); x < viewportRect.maxX(); x += hitTestRegionWidth) {
+        for (auto y = viewportRect.y() + std::floor(hitTestRegionHeight / 2); y < viewportRect.maxY(); y += hitTestRegionHeight) {
+            const FloatPoint hitTestPoint = FloatPoint(x, y);
+            auto [nodes, innerElement] = findNodes(hitTestPoint, true);
+            if (nodes.isEmpty())
+                continue;
+
+            const auto targets = extractTargets(WTFMove(nodes), WTFMove(innerElement), false);
+            targetsList.append(targets);
+        }
+    }
+
+    HashMap<WebCore::ElementIdentifier, HashSet<WebCore::ElementIdentifier>> elementIDToOccludedElementIDs;
+    HashMap<WebCore::ElementIdentifier, Vector<TargetedElementInfo>> elementIDToTargets;
+    for (auto& targets : targetsList) {
+        if (targets.size() <= 0)
+            continue;
+
+        const auto topElementID = targets.first().elementIdentifier;
+        HashSet<WebCore::ElementIdentifier> occludedElementIDsToInsert;
+        bool isFirstElement = true;
+        for (auto& target : targets) {
+            if (isFirstElement) {
+                isFirstElement = false;
+                continue;
+            }
```

Comment:
This is not being used anywhere? Safe to remove?

Reply:
I'm using it to skip adding the first element in targets since that is the key in the dict, so it would create a cycle in the graph. lmk if there's a cleaner way to skip the first element!

Reply:
This is fine! I would've suggested `for (auto& target : std::ranges::drop_view { targets, 2 })` but this is WebCore code that needs to build on platforms that don't have great STL support.

Reply:
Actually why not just do an index/iterator-based loop that starts from the element 1? Might be easier to follow than the `isFirstElement` logic?

```
for (auto index = 1; index < targets.size(); ++target) {
    auto& target = targets.at(index);
    ...
}
```